core-java/src/main/java/eu/neverblink/jelly/core/internal/ProtoDecoderImpl.java

```diff
@@ -0,0 +1,439 @@
+     */
+    public static final class AnyStatementDecoder<TNode, TDatatype> extends ProtoDecoderImpl<TNode, TDatatype> {
+
+        private final RdfHandler.AnyRdfHandler<TNode> protoHandler;
+        private ProtoDecoderImpl<TNode, TDatatype> delegateDecoder = null;
+
+        public AnyStatementDecoder(
+            ProtoDecoderConverter<TNode, TDatatype> converter,
+            RdfHandler.AnyRdfHandler<TNode> protoHandler,
+            RdfStreamOptions supportedOptions
+        ) {
+            super(converter, protoHandler, supportedOptions);
+            this.protoHandler = protoHandler;
+        }
+
+        @Override
+        public RdfStreamOptions getStreamOptions() {
+            if (delegateDecoder != null) {
+                return delegateDecoder.getStreamOptions();
+            }
+
+            return null;
+        }
+
+        @Override
+        public void ingestRow(RdfStreamRow row) {
+            if (row.hasOptions()) {
+                handleOptions(row.getOptions());
+                delegateDecoder.ingestRow(row);
+                return;
+            }
+
+            if (delegateDecoder == null) {
+                throw new RdfProtoDeserializationError("Stream options are not set.");
+            }
+
+            delegateDecoder.ingestRow(row);
+        }
+
+        @Override
+        protected void handleOptions(RdfStreamOptions options) {
+            final var newSupportedOptions = supportedOptions
+                .toBuilder()
+                .setLogicalType(LogicalStreamType.LOGICAL_STREAM_TYPE_UNSPECIFIED)
+                .build();
+
+            checkCompatibility(options, newSupportedOptions);
+            if (delegateDecoder != null) {
+                return;
+            }
+
+            switch (options.getPhysicalType()) {
+                case PHYSICAL_STREAM_TYPE_TRIPLES -> delegateDecoder = new TriplesDecoder<>(
+                    converter,
+                    protoHandler,
+                    options
+                );
+                case PHYSICAL_STREAM_TYPE_QUADS -> delegateDecoder = new QuadsDecoder<>(
+                    converter,
+                    protoHandler,
+                    options
+                );
+                case PHYSICAL_STREAM_TYPE_GRAPHS -> delegateDecoder = new GraphsAsQuadsDecoder<>(
+                    converter,
+                    protoHandler,
+                    options
+                );
+                default -> throw new RdfProtoDeserializationError("Incoming physical stream type is not recognized.");
+            }
+        }
+
+        @Override
+        protected void handleTriple(RdfTriple triple) {
+            delegateDecoder.handleTriple(triple);
+        }
+
+        @Override
+        protected void handleQuad(RdfQuad quad) {
+            delegateDecoder.handleQuad(quad);
+        }
+
+        @Override
+        protected void handleGraphStart(RdfGraphStart graphStart) {
+            delegateDecoder.handleGraphStart(graphStart);
```

Comment:
Don't worry, this will not produce a graphStart event on the handler. The delegate is only GraphsAsQuadsDecoder. In Scala this would be nicely typed, but with Java we can't have nice things.